File: backend/routers/projects.py
import re
from typing import List

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

import models
import schemas
from database import get_db

router = APIRouter(tags=["projects"])

VALID_ROLES = {"character", "scene", "background", "prop", "concept", "reference", "other"}


def slug_to_name(slug: str) -> str:
    return slug.replace("-", " ").title()


def normalize_slug(name: str) -> str:
    slug = name.strip().lower()
    slug = re.sub(r"[^a-z0-9]+", "-", slug)
    slug = slug.strip("-")
    return slug


@router.get("/api/projects", response_model=List[schemas.ProjectInfo])
def list_projects(db: Session = Depends(get_db)):
    tags = db.query(models.Tag).filter(models.Tag.name.like("project:%")).all()

    projects: dict[str, dict] = {}
    for tag in tags:
        parts = tag.name.split(":")
        # Only top-level project tags: "project:<slug>" (exactly 2 parts)
        if len(parts) == 2:
            slug = parts[1]
            count = len(tag.images)
            projects[slug] = {"slug": slug, "name": slug_to_name(slug), "image_count": count}

    return [schemas.ProjectInfo(**p) for p in projects.values()]


@router.get("/api/projects/{slug}", response_model=schemas.ProjectDetail)
def get_project(slug: str, db: Session = Depends(get_db)):
    project_tag = db.query(models.Tag).filter(models.Tag.name == f"project:{slug}").first()

    if not project_tag:
        raise HTTPException(status_code=404, detail="Project not found")

    images_in_project = list(project_tag.images)

    grouped: dict[str, list] = {role: [] for role in VALID_ROLES}

    for img in images_in_project:
        img_tag_names = {t.name for t in img.tags}
        img_roles = [role for role in VALID_ROLES if f"project:{slug}:{role}" in img_tag_names]
        if not img_roles:
            img_roles = ["other"]
        for role in img_roles:
            grouped[role].append(img)

    # Remove empty role buckets
    grouped = {k: v for k, v in grouped.items() if v}

    return schemas.ProjectDetail(
        slug=slug,
        name=slug_to_name(slug),
        image_count=len(images_in_project),
        roles=grouped,
    )


@router.post("/api/images/{image_id}/project", response_model=schemas.Image)
def assign_project(image_id: int, body: schemas.ProjectAssignRequest, db: Session = Depends(get_db)):
    img = db.query(models.Image).filter(models.Image.id == image_id).first()
    if not img:
        raise HTTPException(status_code=404, detail="Image not found")

    slug = normalize_slug(body.project)
    if not slug:
        raise HTTPException(status_code=400, detail="Invalid project name")

    tag_names = [f"project:{slug}"]
    for role in body.roles:
        role = role.strip().lower()
        if role in VALID_ROLES:
            tag_names.append(f"project:{slug}:{role}")

    for tag_name in tag_names:
        tag = db.query(models.Tag).filter(models.Tag.name == tag_name).first()
        if not tag:
            tag = models.Tag(name=tag_name)
            db.add(tag)
            db.flush()
        if tag not in img.tags:
            img.tags.append(tag)

    db.commit()
    db.refresh(img)
    return img


@router.delete("/api/images/{image_id}/project", response_model=schemas.Image)
def remove_project(image_id: int, body: schemas.ProjectRemoveRequest, db: Session = Depends(get_db)):
    img = db.query(models.Image).filter(models.Image.id == image_id).first()
    if not img:
        raise HTTPException(status_code=404, detail="Image not found")

    slug = normalize_slug(body.project)

    if body.roles:
        for role in body.roles:
            role = role.strip().lower()
            if role not in VALID_ROLES:
                continue
            tag = db.query(models.Tag).filter(models.Tag.name == f"project:{slug}:{role}").first()
            if tag and tag in img.tags:
                img.tags.remove(tag)
    else:
        to_remove = [
            t for t in img.tags
            if t.name == f"project:{slug}" or t.name.startswith(f"project:{slug}:")
        ]
        for tag in to_remove:
            img.tags.remove(tag)

    db.commit()
    db.refresh(img)
    return img

File: frontend/src/stores/projects.ts
import { defineStore } from 'pinia'
import axios from 'axios'
import type { Image, ProjectDetail, ProjectInfo } from '../types/api'

interface ProjectsState {
  projects: ProjectInfo[]
  loading: boolean
}

type RoleAssignments = Record<string, string[]>

export const useProjectsStore = defineStore('projects', {
  state: (): ProjectsState => ({
    projects: [],
    loading: false,
  }),

  actions: {
    async fetchProjects(): Promise<void> {
      this.loading = true
      try {
        const res = await axios.get<ProjectInfo[]>('/api/projects')
        this.projects = res.data
      } catch (e) {
        console.error('fetchProjects failed', e)
      } finally {
        this.loading = false
      }
    },

    async assignProject(
      imageId: number,
      project: string,
      roles: RoleAssignments,
    ): Promise<Image> {
      const res = await axios.post<Image>(`/api/images/${imageId}/project`, {
        project,
        roles,
      })
      await this.fetchProjects()
      return res.data
    },

    async removeProject(
      imageId: number,
      project: string,
      roles: RoleAssignments | null = null,
    ): Promise<Image> {
      const body: { project: string; roles?: RoleAssignments } = { project }
      if (roles) body.roles = roles
      const res = await axios.delete<Image>(`/api/images/${imageId}/project`, {
        data: body,
      })
      await this.fetchProjects()
      return res.data
    },

    async fetchProject(slug: string): Promise<ProjectDetail> {
      const res = await axios.get<ProjectDetail>(`/api/projects/${slug}`)
      return res.data
    },
  },
})
